node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright 2017-2022 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	error::{Error, GetOnchainDisputesError},
+	metrics, LOG_TARGET,
+};
+use futures::channel::oneshot;
+use polkadot_node_primitives::{CandidateVotes, DisputeStatus};
+use polkadot_node_subsystem::{messages::DisputeCoordinatorMessage, overseer, ActivatedLeaf};
+use polkadot_primitives::v2::{
+	supermajority_threshold, CandidateHash, DisputeState, DisputeStatement, DisputeStatementSet,
+	Hash, InvalidDisputeStatementKind, MultiDisputeStatementSet, SessionIndex,
+	ValidDisputeStatementKind, ValidatorIndex, ValidatorSignature,
+};
+use std::collections::HashMap;
+
+#[cfg(test)]
+mod tests;
+
+/// The maximum number of disputes Provisioner will include in the inherent data.
+/// Serves as a protection not to flood the Runtime with excessive data.
+pub const MAX_DISPUTE_VOTES_FORWARDED_TO_RUNTIME: usize = 200 * 1_000; // `VALIDATORS COUNT` * `DISPUTES COUNT`
+
+pub async fn select_disputes<Sender>(
+	sender: &mut Sender,
+	metrics: &metrics::Metrics,
+	_leaf: &ActivatedLeaf,
+) -> Result<MultiDisputeStatementSet, Error>
+where
+	Sender: overseer::ProvisionerSenderTrait,
+{
+	// Fetch the onchain disputes. We'll do a prioritisation based on them.
+	let onchain = match get_onchain_disputes(sender, _leaf.hash.clone()).await {
+		Ok(r) => r,
+		Err(e) => {
+			gum::debug!(
+				target: LOG_TARGET,
+				?e,
+				"Can't fetch onchain disputes. Will continue with empty onchain disputes set.",
+			);
+			HashMap::new()
+		},
+	};
+
+	let recent_disputes = request_disputes(sender).await;
+	let partitioned = partition_recent_disputes(recent_disputes, &onchain);
+
+	let result = vote_selection(sender, partitioned, &onchain).await;
+
+	// TODO: sort result?
```

Comment:
In the old implementation disputes were sorted by session index in `fn extend_by_random_subset_without_repetition()`:
```
	// assure sorting stays candid according to session index
	acc.sort_unstable_by(|a, b| a.0.cmp(&b.0));
```
Should we still do this here considering that we prioritise the votes in a different way now?